python/lib/core/dmod/core/common/helper_functions.py

```diff
@@ -174,4 +183,323 @@ def merge_dictionaries(first: typing.Mapping = None, second: typing.Mapping = No
+        10
+        >>> truncate(example_value, 3)
+        10.123
+        >>> truncate(example_value, 5)
+        10.12312
+
+    Args:
+        number: The number to truncate
+        digits: The number of desired digits
+
+    Returns:
+        The floating point number truncated to the specified number of digits
+    """
+    if digits <= 0:
+        return number
+
+    if not isinstance(digits, int):
+        raise TypeError(
+            f"The number {number} cannot be truncated - "
+            f"the number of digits to truncate by must be an integer and received {type(digits)} instead"
+        )
+
+    # We'll use the number 10.23423523423 and the desired number of digits as 3 as an example
+
+    # This gives us "10.23423523423"
+    string_representation = str(number)
+
+    if "." not in string_representation:
+        return number
+
+    # Find the number of current decimal points in order to avoid operations if necessary.
+    # This will yield 11 for our example
+    number_of_decimal_points = len(string_representation.split('.')[1])
+
+    # If the current number of digits is less than or equal to the number of desired digits, we can move on
+    # Our example has 11 decimal points, which is greater than the desired 3, so we continue
+    if number_of_decimal_points <= digits:
+        return number
+
+    # Create a factor that will shift the desired number of decimal points
+    # This will create 1000 for our example
+    adjuster = 10.0 ** digits
+
+    # This will shift the number by the desired number of decimal points
+    # This yields 10234.23523423 for our example
+    adjusted_value = adjuster * number
+
+    # This will get rid of the extra digits
+    # Our example value now becomes 10234.0
+    adjusted_value = math.trunc(adjusted_value)
+
+    # This will shift the value back to the desired number of decimal points
+    # Our example value now becomes 10.324
+    return adjusted_value / adjuster
+
+
+def get_subclasses(base: typing.Type[_CLASS_TYPE]) -> typing.List[typing.Type[_CLASS_TYPE]]:
+    """
+    Gets a collection of all concrete subclasses of the given class in memory
+
+    A subclass that has not been imported will not be returned
+
+    Example:
+        >>> import numpy
+        >>> get_subclasses(float)
+        [numpy.float64]
+
+    Args:
+        base: The base class to get subclasses from
+
+    Returns:
+        All implemented subclasses of a specified types
+    """
+    subclasses = [
+        cls
+        for cls in base.__subclasses__()
+    ]
+
+    concrete_classes = [
+        subclass
+        for subclass in subclasses
+        if not inspect.isabstract(subclass)
+    ]
+
+    for subclass in subclasses:
+        concrete_classes.extend([
+            cls
+            for cls in get_subclasses(subclass)
+            if cls not in concrete_classes
+               and not inspect.isabstract(cls)
+        ])
+
+    return concrete_classes
+
+
+def foreach(
```

Comment:
Is this just for naming?

I think it was originally `each`, but I went with `foreach` since that's a more common term used across programming languages. It is a little more obvious than `map` (and `greedy_map` by extension), so I'd prefer some form of term that explicitly states "I'm calling this for everything in the collection" since `for` logic is more of a 100 level idea with `map` being a 200 level.

Reply:
Yeah, just about naming. I was just being pedantic. I don't think _greedy_ was the word I was looking for when I originally commented, instead I think I was looking for _eager_. Either way, I like your logical hierarchy between `for` and `map`. And to be clear, im totally cool leaving this how it is! With that in mind, the only thing I am debating is the function name because it returns a new sequence rather than mutating the existing collection. I am also biased because that is how [`javascipt`'s `foreach` function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) works.

Connecting that idea back to `map` and `for`, when _I_ think of `map`, I think it will return a new collection that does not modify the original collection (its _pure_; ah yes, the 200 level jargon). Where as `foreach` sounds like it should modify a value in place. Thoughts?

Reply:
That's a good point. I updated the function to update the original collection if it was mutable and default to the old behavior if it isn't.

It doesn't have to return anything - that was mostly an afterthought. `map` should be used for that (I'd think), but if someone is calling a function that happens to return a value, why dump the result immediately?

Reply:
Post-Mortem update: I'm removing the value storage/returning/updating mechanism and renaming this 'on_each' (since it operates "on each").

If you use the example function `p` from the docstring with this function, it will replace the collection `[1, 2, 3]` with `[None, None, None]` in place since `p` doesn't return anything. I can't think of a good way to detect if the value in the collection _should_ be updated either. You can only tell if the `func` function returns a value if the author annotated the result (not 100% of the time), so you can't use that to indicate that a returned `None` means that there shouldn't be an update.